File: typescript/src/domain/Animal.ts
export type AnimalSpecies = "Dog" | "Cat" | "Duck";
export type Emotion = "ok" | "satisfied" | "sad";

const FOOD_REACTIONS: Record<
  AnimalSpecies,
  { food: string; emotion: Emotion; energy: number }
> = {
  Dog: { food: "bone", emotion: "ok", energy: 2 },
  Cat: { food: "fish", emotion: "satisfied", energy: 5 },
  Duck: { food: "shit", emotion: "sad", energy: -10 },
};

const SOUNDS: Record<AnimalSpecies, string> = {
  Dog: "Woof",
  Cat: "Meow",
  Duck: "Quack",
};

export class Animal {
  readonly species: AnimalSpecies;
  private _state: "hungry" | "fed" = "hungry";
  private _emotion: Emotion | null = null;
  private _energy: number = 0;

  constructor(species: AnimalSpecies) {
    this.species = species;
  }

  get state() {
    return this._state;
  }

  get emotion(): Emotion | null {
    return this._emotion;
  }

  get energy(): number {
    return this._energy;
  }

  feed(): void {
    const reaction = FOOD_REACTIONS[this.species];
    this._emotion = reaction.emotion;
    this._energy += reaction.energy;
    this._state = "fed";
  }

  makeSound(): string {
    return SOUNDS[this.species];
  }
}

File: golang/domain/animal.go
package domain

type Species string
type Emotion string

const (
	Dog  Species = "Dog"
	Cat  Species = "Cat"
	Duck Species = "Duck"
)

const (
	EmotionOk        Emotion = "ok"
	EmotionSatisfied Emotion = "satisfied"
	EmotionSad       Emotion = "sad"
)

type foodReaction struct {
	food   string
	emotion Emotion
	energy int
}

var foodReactions = map[Species]foodReaction{
	Dog:  {food: "bone", emotion: EmotionOk, energy: 2},
	Cat:  {food: "fish", emotion: EmotionSatisfied, energy: 5},
	Duck: {food: "shit", emotion: EmotionSad, energy: -10},
}

var sounds = map[Species]string{
	Dog:  "Woof",
	Cat:  "Meow",
	Duck: "Quack",
}

type Animal struct {
	Species Species
	state   string
	emotion *Emotion
	energy  int
}

func NewAnimal(species Species) *Animal {
	return &Animal{
		Species: species,
		state:   "hungry",
	}
}

func (a *Animal) State() string {
	return a.state
}

func (a *Animal) Emotion() *Emotion {
	return a.emotion
}

func (a *Animal) Energy() int {
	return a.energy
}

func (a *Animal) Feed(food string) {
	reaction := foodReactions[a.Species]
	if reaction.food != food {
		return
	}
	a.emotion = &reaction.emotion
	a.energy += reaction.energy
	a.state = "fed"
}

func (a *Animal) MakeSound() string {
	return sounds[a.Species]
}
